node/node.go

```diff
@@ -236,3 +193,19 @@ func (n *Node) EventBus() *lltypes.EventBus {
 func (n *Node) ProxyApp() proxy.AppConns {
 	return n.proxyApp
 }
+
+func newProxyAppValidator(app proxy.AppConns) pubsub.Validator {
+	return func(c context.Context, i peer.ID, m *pubsub.Message) bool {
+		req := abci.RequestCheckTx{
+			Tx:   m.Data,
+			Type: abci.CheckTxType_New,
+		}
+		resp, err := app.Mempool().CheckTxSync(c, req)
```

Comment:
at least in BaseApp, calling CheckTx uses the same state until calling Commit resets it. So, I think calling CheckTx additional times could actually lead to some weird edge cases. Still not sure, but it's something I'm investigating.

Reply:
I think Evan is more worried about any side-effects _in the app_ of calling checkTx multiple times. @tzdybal are you saying that this won't happen anyways bc it gets prevented by the mempool that checkTx gets called multiple times effectively?

Reply:
This PR is currently directly using the proxy app, which doesn't check to see if the tx is in the mempool. @liamsi  is correct, that's what I was worried about.

 As @tzdybal hinted, I think changing the validator to not directly use the app, and instead use the mempool's wrapper around `CheckTx`, is a significant improvement over what is already implemented in this PR. There is no reason to run `CheckTx` on a tx that we've already checked and is submitted to our local mempool, and using the mempool's wrapper around `CheckTx` fixes this.